/**
 * Copyright (c) 2015 - The CM Authors <legal@clickmatcher.com>
 *   All Rights Reserved.
 *
 * This file is CONFIDENTIAL -- Distribution or duplication of this material or
 * the information contained herein is strictly forbidden unless prior written
 * permission is obtained.
 */
import "CodingOptions.proto";
import "Currency.proto";

package cm;

enum ProtoLanguage {
  LANGUAGE_UNKNOWN_LANGUAGE = 0;
  LANGUAGE_DE = 2;
  LANGUAGE_EN = 1;
  LANGUAGE_ES = 3;
  LANGUAGE_FR = 4;
  LANGUAGE_IT = 5;
  LANGUAGE_NL = 6;
  LANGUAGE_PL = 7;
}

enum ProtoPageType {
  PAGETYPE_UNKNOWN_PAGE = 0;
  PAGETYPE_SEARCH_PAGE = 1;
  PAGETYPE_CATALOG_PAGE = 2;
  PAGETYPE_PRODUCT_PAGE = 3;
  PAGETYPE_SHOP_PAGE = 4;
}

enum ProtoDeviceType {
  DEVICETYPE_UNKNOWN_DEVICE = 0;
  DEVICETYPE_DESKTOP = 1;
  DEVICETYPE_PHONE = 2;
  DEVICETYPE_TABLET = 3;
}

/**
 * A JoinedSearchQueryResultItem describes a single item in a search queries
 * result list (i.e. a single "placement").
 */
message JoinedSearchQueryResultItem {

  /**
   * The item id of the result item/product
   */
  optional string item_id = 20
      [(stx.coding) = { maxval: 1024 }];

  /**
   * The position at which this item was shown. Start from 1 and 1 is the first
   * result item (i.e. the topmost lefmost one)
   */
  optional uint32 position = 19
      [(stx.coding) = { maxval: 64 encoding: "BITPACK" }];

  /**
   * True if the result item was subsequently clicked
   */
  optional bool clicked = 15;

  /**
   * True if the result item was "seen" by the user (i.e. it was in the
   * viewport)
   */
  optional bool seen = 65;

  /**
   * If applicable, the result item's shop id (DaWanda user id)
   */
  optional uint32 shop_id = 21
      [(stx.coding) = { maxval: 4294967295 encoding: "LEB128" }];

  /**
   * If applicable, the item's category1/2/3 id (DaWanda e1/2/3)
   */
  optional uint32 category1 = 22
      [(stx.coding) = { maxval: 65535 encoding: "LEB128" }];

  optional uint32 category2 = 23
      [(stx.coding) = { maxval: 65535 encoding: "LEB128" }];

  optional uint32 category3 = 24
      [(stx.coding) = { maxval: 65535 encoding: "LEB128" }];

  /**
   * True if this result item is a a paid result item (advertisment)
   */
  optional bool is_paid_result = 62;

  /**
   * True if this result item is a "recommendation" result (generated from
   * a recommendation engine)
   */
  optional bool is_recommendation = 63;

}


/**
 * A JoinedSearchQuery is added to the session for each search query that
 * the user enters
 */
message JoinedSearchQuery {

  /**
   * The time at which this search query was observer
   */
  optional uint64 time = 18;

  /**
   * The language of the search query (i.e. on which language page/subdomain
   * was the query executed)
   */
  optional ProtoLanguage language = 69;

  /**
   * The "device type" (mobile/table/desktop) for this query. N.B. that the
   * classification is currently based upon screen dimensions which is very
   * imprecise (to the point of being useless)
   */
  optional ProtoDeviceType device_type = 10;

  /**
   * The "page type" (catalog/search/shop/etc) on which this query was executed
   */
  optional ProtoPageType page_type = 11;

  /**
   * The "querty type" (catalog/search/shop/recommendation/etc) as a string
   */
  optional string query_type = 64
      [(stx.coding) = { maxval: 1024 }];

  /**
   * The page number of the query was paginated. Starts from 1 == first page
   */
  optional uint32 page = 68
      [(stx.coding) = { maxval: 100 encoding: "BITPACK" }];

  /**
   * The raw search query string (as it was submitted by the user)
   */
  optional string query_string = 3
      [(stx.coding) = { maxval: 8192 }];

  /**
   * The normalized search query string (tokenized, gargabe removed, lowercase)
   */
  optional string query_string_normalized = 4
      [(stx.coding) = { maxval: 8192 }];

  /**
   * The category1/2/3 filter if this was a fulltext search query with category
   * filter or the catalog category 1/2/3 (DaWanda e1/e2/e3) if this was a
   * catalog view. N.B  that more than one of these can be set at the same
   * time (e.g. for a DaWanda e2 view we will set category1 and category2)
   */
  optional uint32 category1 = 12
      [(stx.coding) = { maxval: 65535 encoding: "LEB128" }];

  optional uint32 category2 = 13
      [(stx.coding) = { maxval: 65535 encoding: "LEB128" }];

  optional uint32 category3 = 14
      [(stx.coding) = { maxval: 65535 encoding: "LEB128" }];

  /**
   * If this was a shop page view / search within a shop the shop id of the
   * shop
   */
  optional uint32 shop_id = 32
      [(stx.coding) = { maxval: 4294967295 encoding: "LEB128" }];

  /**
   * The full result list for this query
   */
  repeated JoinedSearchQueryResultItem result_items = 17;

  /**
   * How many result items were returned for this query
   */
  optional uint32 num_result_items = 5
      [(stx.coding) = { maxval: 250 encoding: "BITPACK" }];

  /**
   * How many of the query result items were subsequently clicked
   */
  optional uint32 num_result_items_clicked = 6
      [(stx.coding) = { maxval: 250 encoding: "BITPACK" }];

  /**
   * How many of the shown result items were ads
   */
  optional uint32 num_ad_impressions = 7
      [(stx.coding) = { maxval: 250 encoding: "BITPACK" }];

  /**
   * How many of the shown ads were clicked
   */
  optional uint32 num_ad_clicks = 8
      [(stx.coding) = { maxval: 250 encoding: "BITPACK" }];

  /**
   * How many cart items resulted from this query (i.e. result items that have
   * been subsequently clicked and added to the cart)
   */
  optional uint32 num_cart_items = 43
      [(stx.coding) = { maxval: 250 encoding: "BITPACK" }];

  /**
   * How many order items resulted from this query (i.e. result items that have
   * been subsequently clicked and bought)
   */
  optional uint32 num_order_items = 44
      [(stx.coding) = { maxval: 250 encoding: "BITPACK" }];

  /**
   * The gross value of all cart items that resultes from this query
   */
  optional uint32 cart_value_eurcents = 45
      [(stx.coding) = { maxval: 16777215 encoding: "LEB128" }];

  /**
   * The gross merchandise value / gross revenue that resulted from this query.
   * N.B. that when attributing revenue to queries we attribute the revenue to
   * all candidates in case of an ambiguity and some revenue can't be attributed
   * to any query. So this particular field will not give the correct total
   * revenue when summed up
   */
  optional uint32 gmv_eurcents = 46
      [(stx.coding) = { maxval: 16777215 encoding: "LEB128" }];

  /**
   * A number indicating in which test group the user that submitted this query
   * was. DaWanda sets this to (1..100]
   */
  optional uint32 ab_test_group = 9
      [(stx.coding) = { maxval: 100 encoding: "BITPACK" }];

  /**
   * A semicolon-separated string of all experiment tags for this query
   */
  optional string experiments = 55
      [(stx.coding) = { maxval: 8192 }];

}


/**
 * A JoinedPageView is added to the session for each single visit/click on
 * an item page (via search or otherwise)
 */
message JoinedPageView {
  /**
   * The time at which the item was visited
   */
  optional uint64 time = 26;

  /**
   * The item id of the item that was visited
   */
  optional string item_id = 27
      [(stx.coding) = { maxval: 1024 }];

  /**
   * If applicable, the item's shop id (DaWanda user id)
   */
  optional uint32 shop_id = 28
      [(stx.coding) = { maxval: 4294967295 encoding: "LEB128" }];

  /**
   * If applicable, the item's category1/2/3 id (DaWanda e1/2/3)
   */
  optional uint32 category1 = 29
      [(stx.coding) = { maxval: 65535 encoding: "LEB128" }];

  optional uint32 category2 = 30
      [(stx.coding) = { maxval: 65535 encoding: "LEB128" }];

  optional uint32 category3 = 31
      [(stx.coding) = { maxval: 65535 encoding: "LEB128" }];


}


/**
 * A JoinedCartItem is added to the session for every single cart item that
 * was observed (added/bought/etc)
 */
message JoinedCartItem {

  /**
   * The time at which this item was last seen in the users cart
   */
  optional uint64 time = 33;

  /**
   * The item id of the item (=product) that was added to the cart
   */
  optional string item_id = 34
      [(stx.coding) = { maxval: 1024 }];

  /**
   * The quantity of this cart item. N.B. that we always store the largest
   * quantity obeerved
   */
  optional uint32 quantity = 39
      [(stx.coding) = { maxval: 65535 encoding: "LEB128" }];

  /**
   * The price in cents and currency for a single instance of the item
   */
  optional uint32 price_cents = 40
      [(stx.coding) = { maxval: 4294967295 encoding: "LEB128" }];

  optional stx.Currency currency = 41;

  /**
   * The most advanced cart step we have seen this item in. Note that smaller
   * value means more advanced -- step 1 means "transaction confirmed" and
   * implied the item was bought
   */
  optional uint32 checkout_step = 42
      [(stx.coding) = { maxval: 32 encoding: "BITPACK" }];

  /**
   * If applicable, the product's/item's shop id (DaWanda user id)
   */
  optional uint32 shop_id = 35
      [(stx.coding) = { maxval: 4294967295 encoding: "LEB128" }];

  /**
   * If applicable, the product's/item's category1/2/3 id (DaWanda e1/2/3)
   */
  optional uint32 category1 = 36
      [(stx.coding) = { maxval: 65535 encoding: "LEB128" }];

  optional uint32 category2 = 37
      [(stx.coding) = { maxval: 65535 encoding: "LEB128" }];

  optional uint32 category3 = 38
      [(stx.coding) = { maxval: 65535 encoding: "LEB128" }];

}


/**
 * A JoinedSessions contains all information of a single user journey. I.e.
 * all events that were submitted for a specific user id within a relatively
 * short (~90m) timeframe.
 */
message JoinedSession {

  /**
   * Deep Analytics session id
   */
  optional string session_id = 66;

  /**
   * Nominal session time (emit time)
   */
  optional uint64 time = 67;

  /**
   * For which customer did we record this session?
   */
  optional string customer = 60
      [(stx.coding) = { maxval: 250 }];

  /**
   * The customer's session id
   */
  optional string customer_session_id = 61
      [(stx.coding) = { maxval: 4096 }];

  /**
   * The time at which this search query was first/last seen
   */
  optional uint32 first_seen_time = 53;
  optional uint32 last_seen_time = 54;

  optional string referrer_url = 57
      [(stx.coding) = { maxval: 1024 }];

  optional string referrer_campaign = 58
      [(stx.coding) = { maxval: 4096 }];

  optional string referrer_name = 59
      [(stx.coding) = { maxval: 255 }];

  /**
   * How many unique cart items have we seen in this session
   */
  optional uint32 num_cart_items = 47
      [(stx.coding) = { maxval: 250 encoding: "BITPACK" }];

  /**
   * How many unique items have been bought in this session. value >1 implies
   * that there was a transaction in this session
   */
  optional uint32 num_order_items = 48
      [(stx.coding) = { maxval: 250 encoding: "BITPACK" }];

  /**
   * The gross value of all cart items in this session in euro cents
   */
  optional uint32 cart_value_eurcents = 49
      [(stx.coding) = { maxval: 16777215 encoding: "LEB128" }];

  /**
   * The gross merchandise value of all transactions/sales in this session in
   * euro cents
   */
  optional uint32 gmv_eurcents = 50
      [(stx.coding) = { maxval: 16777215 encoding: "LEB128" }];

  /**
   * A number indicating in which test group this session was. DaWanda sets this
   * to (1..100]
   */
  optional uint32 ab_test_group = 51
      [(stx.coding) = { maxval: 100 encoding: "BITPACK" }];

  /**
   * A semicolon-separated string of all experiment tags for this session
   */
  optional string experiments = 56
      [(stx.coding) = { maxval: 8192 }];

  /**
   * A list of all search queries associated with this session
   */
  repeated JoinedSearchQuery search_queries = 16;

  /**
   * A list of all item visits associated with this session
   */
  repeated JoinedPageView page_views = 25;

  /**
   * A list of all cart items associated with this session
   */
  repeated JoinedCartItem cart_items = 52;

}


message DefaultSessionAttributes {

  /**
   * The customer's session id
   */
  optional string customer_session_id = 61
      [(stx.coding) = { maxval: 4096 }];

  optional string referrer_url = 57
      [(stx.coding) = { maxval: 1024 }];

  optional string referrer_campaign = 58
      [(stx.coding) = { maxval: 4096 }];

  optional string referrer_name = 59
      [(stx.coding) = { maxval: 255 }];

  /**
   * How many unique cart items have we seen in this session
   */
  optional uint32 num_cart_items = 47
      [(stx.coding) = { maxval: 250 encoding: "BITPACK" }];

  /**
   * How many unique items have been bought in this session. value >1 implies
   * that there was a transaction in this session
   */
  optional uint32 num_order_items = 48
      [(stx.coding) = { maxval: 250 encoding: "BITPACK" }];

  /**
   * The gross value of all cart items in this session in euro cents
   */
  optional uint32 cart_value_eurcents = 49
      [(stx.coding) = { maxval: 16777215 encoding: "LEB128" }];

  /**
   * The gross merchandise value of all transactions/sales in this session in
   * euro cents
   */
  optional uint32 gmv_eurcents = 50
      [(stx.coding) = { maxval: 16777215 encoding: "LEB128" }];

  /**
   * A number indicating in which test group this session was. DaWanda sets this
   * to (1..100]
   */
  optional uint32 ab_test_group = 51
      [(stx.coding) = { maxval: 100 encoding: "BITPACK" }];

};

message NewSessionEvents {
  
  /**
   * A list of all search queries associated with this session
   */
  repeated JoinedSearchQuery search_query = 16;

  /**
   * A list of all item visits associated with this session
   */
  repeated JoinedPageView page_view = 25;

  /**
   * A list of all cart items associated with this session
   */
  repeated JoinedCartItem cart_items = 52;

};

message NewJoinedSession {

  /**
   * Deep Analytics session id
   */
  optional string session_id = 70;

  /**
   * Nominal session time (emit time)
   */
  optional uint64 time = 71;

  /**
   * The time at which this search query was first/last seen
   */
  optional uint64 first_seen_time = 53;
  optional uint64 last_seen_time = 54;

  optional DefaultSessionAttributes attr = 1;
  optional NewSessionEvents event = 2;

};
